src/model/sidebar.ml

```diff
@@ -0,0 +1,150 @@
+        else (id, payload) :: acc)
+      dir_content.leafs []
+
+  let dir_payload ((parent_id, _) as dir) =
+    let index_id =
+      Paths.Identifier.Mk.leaf_page (parent_id, Names.PageName.make_std "index")
+    in
+    match find_leaf dir index_id with
+    | Some payload -> Some (payload, index_id)
+    | None -> None
+
+  let dirs (_, dir_content) =
+    CPH.fold (fun id payload acc -> (id, payload) :: acc) dir_content.dirs []
+
+  let contents ((dir_id, _) as dir) =
+    let children_order =
+      match dir_payload dir with
+      | Some ({ children_order; _ }, _) -> children_order
+      | None -> None
+    in
+    let children_order =
+      match children_order with
+      | None ->
+          let contents =
+            let leafs =
+              leafs dir
+              |> List.map (fun (id, payload) -> ((id :> Page.t), Entry payload))
+            in
+            let dirs =
+              dirs dir
+              |> List.map (fun (id, payload) -> ((id :> Page.t), Dir payload))
+            in
+            leafs @ dirs
+          in
+          List.sort
+            (fun (x, _) (y, _) ->
+              String.compare (Paths.Identifier.name x) (Paths.Identifier.name y))
+            contents
+      | Some ch ->
+          let open OptionMonad in
+          List.filter_map
+            (fun c ->
+              let id =
+                match c with
+                | Frontmatter.Page name ->
+                    Mk.leaf_page (dir_id, Names.PageName.make_std name)
+                | Dir name -> Mk.page (dir_id, Names.PageName.make_std name)
+              in
+              find dir id >>= fun content -> Some (id, content))
+            ch
+    in
+    children_order
+
+  let dir_payload ((parent_id, _) as dir) =
+    let index_id =
+      Paths.Identifier.Mk.leaf_page (parent_id, Names.PageName.make_std "index")
+    in
+    match find_leaf dir index_id with
+    | Some payload -> Some (payload.title, index_id)
+    | None -> None
+
+  let rec get_or_create (dir : t) (id : container_page) : t =
+    let _, { dirs = parent_dirs; _ } =
+      match get_parent id with
+      | Some parent -> get_or_create dir parent
+      | None -> dir
+    in
+    let current_item =
+      try Some (CPH.find parent_dirs id) with Not_found -> None
+    in
+    match current_item with
+    | Some item -> item
+    | None ->
+        let new_ = empty_t (Some id) in
+        CPH.add parent_dirs id new_;
+        new_
+
+  let add (dir : t) ((id : leaf_page), title, children_order) =
+    let _, dir_content =
+      match get_parent id with
+      | Some parent -> get_or_create dir parent
+      | None -> dir
+    in
+    LPH.replace dir_content.leafs id { title; children_order }
+
+  let of_list l =
+    let dir = empty_t None in
+    List.iter (add dir) l;
+    dir
+end
+
+type toc = PageToc.t
+
+type library = { name : string; units : Paths.Identifier.RootModule.t list }
+
+type page_hierarchy = { hierarchy_name : string; pages : toc }
```

Comment:
I didn't expect the page hierarchy to have a name and be an extra level on top of a `toc`. What if this hierarchy contains an `index.mld` and a children order ? Is here an extra indentation with the name repeated ?

Reply:
This "hierarchy_name" corresponds to the name of the library's package: the one passed in the `-P` option of `odoc compile-index`.
It is currently [turned](https://github.com/ocaml/odoc/blob/master/src/document/sidebar.ml#L126) into a `"{hierarchy_name}'s Pages"` title:

```
# Package1's Pages

- The best way to do foo
  - Guide
  - example
  - tutorial 

# Package2's Pages

- Blazingly fast bar
  - Guide
  - example
  - tutorial 

# Library1's modules

- Foo
- Bar
```

We could possibly remove this `hierarchy_name`, but I would argue that it is better done in another PR (this PR is not introducing the behavior).

(Maybe:
```
# Pages

- The best way to do foo
  - Guide
  - example
  - tutorial 
- Blazingly fast bar
  - Guide
  - example
  - tutorial 

# Libraries

- library1
  - Foo
  - Bar
```
)

Reply:
Makes total sense! I think the hierarchy name is something we'd want in the breadcrumbs and so for clarity, we should also have it in the sidebar.